portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -41,15 +44,23 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    int numComments = getNumComments(request);
-    if (numComments == -1) {
-      numComments = DEFAULT_MESSAGES;
+    UserService userService = UserServiceFactory.getUserService();
+
+    List<String> messages;
+    if (userService.isUserLoggedIn()) {
+      int numComments = getNumComments(request);
+      if (numComments == -1) {
+        numComments = DEFAULT_MESSAGES;
+      }
+
+      numComments = Math.min(numComments, MAX_MESSAGES);
+      messages = getCommentEntities(datastore, numComments).stream().map(
+          entity -> entity.getProperty("text").toString()).collect(Collectors.toList());
+    } else {
+      messages = new ArrayList<>();
+      messages.add("You must be logged in to see comments.");
```

Comment:
This should be a 403 error code with the message about being logged in. Then, your frontend can handle that appropriately. Remember: Pretend someone else is writing a command line tool that talks directly to your servlet.

Reply:
Funny that you mentioned this. Something didn't sit right with me about this one, so I changed it to a 403 in [PR#14](https://github.com/josh8551021/step/pull/14). Not sure if you saw that. I could, however, add an error to the message explaining why it was forbidden. I'll use this comment to inform my other pull request.